src/storage.ts

```diff
@@ -1,13 +1,50 @@
-type IStorage = {
-    count: number;
-};
-
-const defaultStorage: IStorage = {
-    count: 0,
-};
-
-export const storage = {
-    get: (): Promise<IStorage> =>
-        chrome.storage.sync.get(defaultStorage) as Promise<IStorage>,
-    set: (value: IStorage): Promise<void> => chrome.storage.sync.set(value),
-};
+import { writable, type Writable } from 'svelte/store';
+
+/**
+ * Creates a persistent Svelte store backed by Chrome's sync storage.
+ * @template T The type of the store's value
+ * @param key The key to use in Chrome's storage
+ * @param initialValue The initial value of the store
+ * @returns A writable Svelte store
+ */
+export function persistentStore<T>(key: string, initialValue: T): Writable<T> {
+    const store = writable(initialValue);
+    let isUpdatingFromChrome = false;
+    let isUpdatingFromStore = false;
```

Comment:
I'm a little confused by these flags and the possible order of calling subscriptions 😅 

What do you think about this option like the package [svelte-persisted-store](https://github.com/joshnuss/svelte-persisted-store/blob/master/index.ts#L118)? We'll just return our implementation of the `Writable<t>` interface, which we'll persist in the chrome storage.

```ts
export function persistentStore<T>(key: string, initialValue: T): Writable<T> {
    const store = writable(initialValue);

    function updateStorage(value: T): void {
        chrome.storage.sync.set({ [key]: value });
    }

    function watchChrome() {
        chrome.storage.sync.onChanged.addListener((changes) => {
            if (!Object.hasOwn(changes, key)) return;
            store.set(changes[key].newValue);
        });
    }

    // Initialize the store with the value from Chrome storage
    chrome.storage.sync.get(key).then((result) => {
        let value = Object.hasOwn(result, key) ? result[key] : initialValue;
        if (!Object.hasOwn(result, key)) {
            console.log(
                `Persistent store: couldn't find key [${key}] in chrome storage. Default to initial value [${initialValue}]`
            );
        }
        store.set(value);
        watchChrome();
    });

    return {
        set(this: void, value: T): void {
            store.set(value);
            updateStorage(value);
        },
        update(this: void, updater: Updater<T>): void {
            return store.update((prev: T): T => {
                const value = updater(prev);
                updateStorage(value);
                return value;
            });
        },
        subscribe: store.subscribe,
    };
}
```

Reply:
Thus, when updating from chrome storage, we get the following chain:

`chrome.storage.subscribe` -> `store.set`

and when updating from the extension interface we get a chain:

`store.set(value)`, `chrome.storage.sync.set` -> `chrome.storage.subscribe` -> `store.set`

which in theory shouldn't cause a looped update.

Reply:
Yeah, I actually had to experiment with the order of calls quite a bit to reach my desired outcome: a single update per type.
There's a potential problem with the chain you suggest: Svelte stores are synchronous, while Chrome storage is not. So any Svelte piece of code that is subscribed to such a construct would receive the changes twice. Which could lead to flickering on screen or, in some cases, even performing whole actions or re-rendering of components.

Reply:
> There's a potential problem with the chain you suggest: Svelte stores are synchronous, while Chrome storage is not.

But it seems to me that we also cannot guarantee that the storage set promise:

```ts
chrome.storage.sync.set({ [key]: value }).then(() => {
    isUpdatingFromStore = false;
});
```

will be called later than:

```ts
chrome.storage.onChanged.addListener((changes, namespace) => {
```

Which could potentially lead not only to repeated re-rendering, but also to an endless cycle of setting stores. I don't think we should rely on a precise order that can be changed at any moment :)

> Which could lead to flickering on screen or, in some cases, even performing whole actions or re-rendering of components.

As far as I know, Svelte has a shallow-equal mechanism for checking before changing a real DOM and if the Svelte-store settings are set 2 times to the same value, then the second value will not lead to redrawing.

Reply:
What else do you think about this option? Add another parameter to `persistentStore` that will check for equality of states. By default, it will shallow compare the value, which can be overridden in case of complex states with objects.

```ts
function shallowEqual<T>(value1: T, value2: T) {
    return value1 === value2;
}

export function persistentStore<T>(
    key: string,
    initialValue: T,
    areStatesEqual: (value1: T, value2: T) => boolean = shallowEqual
): Writable<T> {

...

// Initialize the store with the value from Chrome storage
chrome.storage.sync.get(key).then((result) => {
    let value = Object.hasOwn(result, key) ? result[key] : initialValue;

    if (!Object.hasOwn(result, key)) {
        console.log(
            `Persistent store: couldn't find key [${key}] in chrome storage. Default to initial value [${initialValue}]`
        );
    }

    if (!areStatesEqual(value, get(store))) {
        store.set(value);
    }
});
```

It seems that we are solving both problems:

1. a single update per type
2. stores are not linked to each other by subscription, which eliminates looping

Reply:
> But it seems to me that we also cannot guarantee that the storage set promise:
> 
> ```ts
> chrome.storage.sync.set({ [key]: value }).then(() => {
>     isUpdatingFromStore = false;
> });
> ```
> 
> will be called later than:
> 
> ```ts
> chrome.storage.onChanged.addListener((changes, namespace) => {
> ```
> (...)
> 
> As far as I know, Svelte has a shallow-equal mechanism for checking before changing a real DOM and if the Svelte-store settings are set 2 times to the same value, then the second value will not lead to redrawing.

Sorry, I don't understand how the first case can happen. Could you write up a concrete example? You can revert the `debug messages` commit to play around with things and see the updates unfold in realtime.

To your second comment: yes, Svelte will not update twice to the same value, but that wasn't what I meant. Imagine the following scenario:
```typescript
# Component.svelte
import { count } from "../storage";

# In storage, `count` is first initialized with 0
$count = 1;
$count = 2;
```
In your proposed solution, the order of calls will be:
```typescript
store.set(1);
updateStorage(1); # network call, takes ~50ms
store.set(2);
updateStorage(2); # network call, takes ~50ms
# ...network call returns, we have the chrome.storage.sync listener:
store.set(1);
# ...network call returns, we have the chrome.storage.sync listener:
store.set(2);
```
That's what I mean when I say flickering or re-rendering. The end result will most likely set the correct value, yes, but instead of having a smooth `0 -> 1 -> 2` you'll get `0 -> 1 -> 2 -> 1 -> 2`.

Reply:
I understand, thanks for the great example! But I still think it's not quite right to rely on the order of calling asynchronous operations.

Is there any extension specification that guarantees this order of calling asynchronous callbacks? In the example, we consider setting a new state from a component.

```ts
function watchStore() {
    store.subscribe((value) => {
        // Prevent circular updates
        if (isUpdatingFromChrome) return;
        isUpdatingFromStore = true; // 1 <---------------------------------------
        chrome.storage.sync.set({ [key]: value }).then(() => {
            isUpdatingFromStore = false; // 3 <---------------------------------------
        });
    });
}

function watchChrome() {
    chrome.storage.onChanged.addListener((changes, namespace) => { // 2 <---------------------------------------
        // Prevent circular updates
        if (isUpdatingFromStore || namespace !== "sync" || !Object.hasOwn(changes, key)) return;
        isUpdatingFromChrome = true;
        store.set(changes[key].newValue);
        isUpdatingFromChrome = false;
    });
}
```

It seems that at any moment `chrome.storage.onChanged.addListener` can, for example, become a microtask and be executed before the promise (also a microtask) from `chrome.storage.sync.set`:

```ts
function watchStore() {
    store.subscribe((value) => {
        // Prevent circular updates
        if (isUpdatingFromChrome) return;
        isUpdatingFromStore = true; // 1 <---------------------------------------
        chrome.storage.sync.set({ [key]: value }).then(() => {
            isUpdatingFromStore = false; // 2 <---------------------------------------
        });
    });
}

function watchChrome() {
    chrome.storage.onChanged.addListener((changes, namespace) => { // 3 <---------------------------------------
        // Prevent circular updates
        if (isUpdatingFromStore || namespace !== "sync" || !Object.hasOwn(changes, key)) return;
        isUpdatingFromChrome = true;
        store.set(changes[key].newValue);
        isUpdatingFromChrome = false;
    });
}
```

And thus we will get an infinity update cycle. I would like to find some more reliable way that will not be tied to the order of calling callbacks from the Svelte store and the Chrome store.

Reply:
https://lapcatsoftware.com/articles/works.html

For example, as I understand it, in Safari the order of calling is not guaranteed:

> In Safari the storage is multi-threaded, database backed — truly async. Given the async nature of all of these calls, I don't think FIFO should be guaranteed.

And I'm also not sure that it is guaranteed in Chrome or Firefox and may be changed at any time 🤔

Reply:
Sorry, still not sure how that (second) example could happen :-)
Note that 1 -> 2 cannot happen before 3 happens, because the order is a. `chrome.storage.sync.set`, b. `listener`, c. `then(() =>`. (c) can never come before (b).
I do recommend to put in the debug statements and then just write some code that sets the svelte store and/or the chrome storage, and see how it plays out. If you can find a code example that ends up with the wrong result or an actual endless loop, we could try to tackle that.

Reply:
I can't find an example with an error. Yes, everything works as it should in Chrome now. But I'm still afraid that it works only today or in certain cases. In my experience, you shouldn't rely on the order of asynchronous calls - this often leads to difficult to reproduce bugs 😅  And I also can't find evidence that the order will always be like this and, for example, after the next Chrome update, the order won't change.

Can you share a link or thoughts on why the order will always be like this and never different?)

```ts
chrome.storage.sync.set({ [key]: value }) // 1
    .then(() => {
        // 3
    });

chrome.storage.onChanged.addListener((changes, namespace) => {
    // 2
});
```

Reply:
<img width="597" alt="image" src="https://github.com/user-attachments/assets/b3d66f0e-914d-4717-aeb3-f2229c308ffa">

And the most interesting thing is if you try this code in the extension service worker, the order will be different 😱

Reply:
Ah, I see the source of confusion now. Note the difference between our case (a listener) and the link you provided: they looked at `.set()` in conjunction with `.get()`, while here it's `.set()` in conjunction with `addListener()`.
You're right that we shouldn't rely on a (pseudo-?) async mechanism firing order. I suppose that in Chrome (and FF?) the response back from the sync mechanism fires off two events: the listener event and resolving the `set().then()` promise, and in that order. It does make some logical sense to me.
But I agree that if they ever decided to change their backend implementation to be truly asynchronous and multithreaded, like for Safari, it is conceivable that the `set()` will resolve before the listener event is fired.

Hmm. I'm honestly not sure how something like that can be solved. Intuitively I'm thinking of managing a semaphore-like system to keep track of updates, but... honestly that just feels like too much pain for something this simple.
Open to suggestions.

Reply:
> Intuitively I'm thinking of managing a semaphore-like system to keep track of updates, but... honestly that just feels like too much pain for something this simple.

Yes, I absolutely agree 🙂